### PR TITLE
#1815 - Ignore restrictions for specific disbursement(s) within an assessment

### PR DESCRIPTION
**Pull request by andrewsignori-aot**
- Created the `restrictionBypassesResolutionStep` to resolve active bypasses when needed as the last step in the e-Cert generation process for part-time and full-time. 
- Changed the `getEligibleDisbursements` query adding the information from the active bypasses and the restriction bypassed.
- Method `getRestrictionByActionType` was changed to use the effective restrictions (active restrictions not bypassed) instead of the active restrictions.
- The BCLM step is ignored at the beginning of its execution. If the BCLM restriction is in place the method would not be executed (logic already in place). The condition based on the bypass was added either way to ensure that whatever happens to the BCLM logic the bypass will guarantee that the store step will be ignored.
- A log was created to list all the active restrictions known during the e-Cert calculation. The log records the restriction code and the student restriction ID for easier troubleshooting.
- Created factories for the new table `ApplicationRestrictionBypass`.

#### Full-time
- Should generate BC awards amounts with no restriction deductions and resolve the restriction bypass when a student has an active 'BCLM' restriction and it is bypassed with behavior 'Next disbursement only'.
- Should have the e-Cert generated for a full-time application and the bypass active when a student has an active 'Stop full time disbursement' restriction and it is bypassed with behavior 'All disbursements'.
- Should prevent an e-Cert generation and keep the bypass active when multiple 'Stop full time disbursement' restrictions exist and only one is bypassed and it is bypassed with behavior 'Next disbursement only'.

#### Part-time
- Should have the e-Cert generated for a part-time application and the bypass resolved when a student has an active 'Stop part time disbursement' restriction and it is bypassed with behavior 'Next disbursement only'.
- Should prevent an e-Cert generation and keep the bypass active when multiple 'Stop part time disbursement' restrictions exist and only one is bypassed and it is bypassed with behavior 'Next disbursement only'.

### Log sample without the bypass

image

### Log sample with a bypass in place.

image

### Noted generated when a bypass is resolved by the system

AC was adjusted later to include the application number.
image

### CODE REVIEW

File: sims.code-workspace [hunk 1]
```diff
@@ -91,6 +91,7 @@
     "cSpell.words": [
       "AEST",
       "bcag",
+      "BCLM",
```
Comment (sh16011993) [inline sims.code-workspace:94]: 👍

**Comment by dheepak-aot** [comment]
The PR Title can end with `within an application`?

**Comment by andrewsignori-aot** [comment]
> The PR Title can end with `within an application`?

It just follows the ticket title.

File: sources/packages/backend/libs/integrations/src/services/disbursement-schedule/disbursement-schedule.models.ts [hunk 2]
```diff
@@ -215,7 +245,12 @@ export class EligibleECertDisbursement {
     public readonly maxLifetimeBCLoanAmount: number,
     public readonly disabilityDetails: DisabilityDetails,
     readonly restrictions: StudentActiveRestriction[],
-  ) {}
+    readonly restrictionBypass: ApplicationActiveRestrictionBypass[],
```
Comment (dheepak-aot) [inline sources/packages/backend/libs/integrations/src/services/disbursement-schedule/disbursement-schedule.models.ts:248]: It can be declared as private as it is not intended to be accessed outside right?

Same for 247, but that is not part of the PR.

Comment (andrewsignori-aot) [reply sources/packages/backend/libs/integrations/src/services/disbursement-schedule/disbursement-schedule.models.ts]: Changed.

File: sources/packages/backend/libs/integrations/src/services/disbursement-schedule/e-cert-processing-steps/e-cert-steps-utils.ts [hunk 3]
```diff
@@ -21,8 +23,31 @@ export function getRestrictionByActionType(
...
+ * Adds to the log the lists of all restriction bypasses active.
+ * @param bypasses active bypasses.
+ * @param log log to receive the list.
+ */
+export function logActiveRestrictionsBypasses(
+  bypasses: ReadonlyArray<ApplicationActiveRestrictionBypass>,
+  log: ProcessSummary,
+): void {
+  if (!bypasses.length) {
+    log.info("There are not active restriction bypasses.");
```
Comment (dheepak-aot) [inline sources/packages/backend/libs/integrations/src/services/disbursement-schedule/e-cert-processing-steps/e-cert-steps-utils.ts:41]: no active restriction bypasses.?

File: sources/packages/backend/libs/integrations/src/services/disbursement-schedule/e-cert-processing-steps/e-cert-steps-utils.ts [hunk 4]
```diff
@@ -21,8 +23,31 @@ export function getRestrictionByActionType(
...
+    .getEffectiveRestrictions()
+    .find((restriction) => restriction.actions.includes(actionType));
+}
+
+/**
+ * Adds to the log the lists of all restriction bypasses active.
+ * @param bypasses active bypasses.
+ * @param log log to receive the list.
+ */
+export function logActiveRestrictionsBypasses(
```
Comment (dheepak-aot) [inline sources/packages/backend/libs/integrations/src/services/disbursement-schedule/e-cert-processing-steps/e-cert-steps-utils.ts:36]: 👍

File: sources/packages/backend/libs/integrations/src/services/disbursement-schedule/e-cert-processing-steps/restriction-bypasses-resolution-step.ts [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+    }
+    // Remove the bypasses.
+    const applicationRestrictionBypassRepo = entityManager.getRepository(
+      ApplicationRestrictionBypass,
+    );
+    // Create the student note to be shared for all bypasses being removed.
+    const studentNote = await this.noteSharedService.createStudentNote(
+      eCertDisbursement.studentId,
+      NoteType.Application,
+      `Automatically removing bypass from application number ${eCertDisbursement.applicationNumber} after the first e-Cert was generated.`,
```
Comment (dheepak-aot) [inline sources/packages/backend/libs/integrations/src/services/disbursement-schedule/e-cert-processing-steps/restriction-bypasses-resolution-step.ts:54]: after the first e-cert with bypass is generated ?

File: sources/packages/backend/libs/integrations/src/services/disbursement-schedule/e-cert-processing-steps/restriction-bypasses-resolution-step.ts [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+    private readonly systemUsersService: SystemUsersService,
+  ) {}
+
+  /**
+   * Remove restrictions active bypasses configured as {@link RestrictionBypassBehaviors.NextDisbursementOnly}.
+   * @param eCertDisbursement eligible disbursement to be potentially added to an e-Cert.
+   * @param entityManager used to execute the commands in the same transaction.
+   * @param log cumulative log summary.
+   */
+  async executeStep(
```
Comment (dheepak-aot) [inline sources/packages/backend/libs/integrations/src/services/disbursement-schedule/e-cert-processing-steps/restriction-bypasses-resolution-step.ts:30]: returns comment.

File: sources/packages/backend/libs/integrations/src/services/disbursement-schedule/e-cert-processing-steps/restriction-bypasses-resolution-step.ts (hunk 5 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/libs/integrations/src/services/disbursement-schedule/e-cert-processing-steps/restriction-bypasses-resolution-step.ts:54]: If it would be a request to change the content, I would not see it as an improvement.
I would prefer to leave as it is unless there is a strong desired to have it changed.

File: sources/packages/backend/libs/integrations/src/services/disbursement-schedule/e-cert-processing-steps/e-cert-steps-utils.ts (hunk 3 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/libs/integrations/src/services/disbursement-schedule/e-cert-processing-steps/e-cert-steps-utils.ts]: Fixed.

File: sources/packages/backend/libs/integrations/src/services/disbursement-schedule/e-cert-processing-steps/restriction-bypasses-resolution-step.ts (hunk 6 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/libs/integrations/src/services/disbursement-schedule/e-cert-processing-steps/restriction-bypasses-resolution-step.ts:31]: Added here and in the base class.

File: sources/packages/backend/libs/integrations/src/services/disbursement-schedule/e-cert-processing-steps/restriction-bypasses-resolution-step.ts [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+    entityManager: EntityManager,
+    log: ProcessSummary,
+  ): Promise<boolean> {
+    log.info("Checking if restrictions bypasses should be removed.");
+    const nextDisbursementOnlyBypasses =
+      eCertDisbursement.activeRestrictionBypasses.filter(
+        (bypass) =>
+          bypass.bypassBehavior ===
+          RestrictionBypassBehaviors.NextDisbursementOnly,
+      );
```
Comment (dheepak-aot) [inline sources/packages/backend/libs/integrations/src/services/disbursement-schedule/e-cert-processing-steps/restriction-bypasses-resolution-step.ts:42]: Let me know if I am missing something here, are we checking if the bypass is relevant to the e-cert. Or is it even required to check that.

for e.g. Consider there is a e-cert for part time which on the completion of all it's steps can set a bypass applied for a restriction which has stop full-time disbursement to inactive.

File: sources/packages/backend/libs/test-utils/src/factories/application-restriction-bypass.ts [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+  bypass.application = relations.application;
+  bypass.studentRestriction = relations.studentRestriction;
+  bypass.bypassBehavior = RestrictionBypassBehaviors.AllDisbursements;
+  bypass.isActive = false;
+  bypass.creationNote = relations.creationNote;
+  bypass.bypassCreatedBy = relations.bypassCreatedBy;
+  bypass.bypassCreatedDate = now;
+  bypass.removalNote = undefined;
+  bypass.bypassRemovedBy = undefined;
+  bypass.bypassRemovedDate = undefined;
```
Comment (dheepak-aot) [inline sources/packages/backend/libs/test-utils/src/factories/application-restriction-bypass.ts:46]: Is it explicitly required to be set undefined here?

File: sources/packages/backend/apps/queue-consumers/src/processors/schedulers/esdc-integration/ecert-integration/_tests_/ecert-full-time-process-integration.scheduler.e2e-spec.ts [hunk 9]
```diff
@@ -696,7 +708,7 @@ describe(
       ]);
       const [disbursement] =
         application.currentAssessment.disbursementSchedules;
-      const isScheduleSent = await db.disbursementSchedule.exist({
+      const isScheduleSent = await db.disbursementSchedule.exists({
```
Comment (dheepak-aot) [inline sources/packages/backend/apps/queue-consumers/src/processors/schedulers/esdc-integration/ecert-integration/_tests_/ecert-full-time-process-integration.scheduler.e2e-spec.ts:711]: 👍

File: sources/packages/backend/libs/integrations/src/services/disbursement-schedule/e-cert-processing-steps/restriction-bypasses-resolution-step.ts (hunk 7 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/libs/integrations/src/services/disbursement-schedule/e-cert-processing-steps/restriction-bypasses-resolution-step.ts:42]: I am not sure if I follow the question. Trying to explain the process next.
1 - `activeRestrictionBypasses` contains all active bypasses associated with the application and they belong to that application context only.
2 - At the end of all steps, any active bypass with the behavior `NextDisbursementOnly` should be removed.
So, we select all active bypasses to be removed and update them.
Please keep in mind that bypasses are applied per application and offering intensity.
Hope that helps, otherwise, we can have a call. Please let me know.

File: sources/packages/backend/libs/test-utils/src/factories/application-restriction-bypass.ts (hunk 8 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/libs/test-utils/src/factories/application-restriction-bypass.ts:46]: Nope, left here to help devs in the future willing to configure the other values.

File: sources/packages/backend/apps/queue-consumers/src/processors/schedulers/esdc-integration/ecert-integration/_tests_/ecert-full-time-process-integration.scheduler.e2e-spec.ts [hunk 10]
```diff
@@ -775,5 +787,372 @@ describe(
...
+              restrictionAmountSubtracted: null,
+            },
+            {
+              valueCode: "BCAG",
+              effectiveAmount: 750,
+              restrictionAmountSubtracted: null,
+            },
+          ]),
+        });
+        // Validate if the restriction was resolved.
```
Comment (dheepak-aot) [inline sources/packages/backend/apps/queue-consumers/src/processors/schedulers/esdc-integration/ecert-integration/_tests_/ecert-full-time-process-integration.scheduler.e2e-spec.ts:925]: Validate if bypass was removed?

File: sources/packages/backend/libs/integrations/src/services/disbursement-schedule/e-cert-processing-steps/restriction-bypasses-resolution-step.ts (hunk 7 above)
Comment (dheepak-aot) [reply sources/packages/backend/libs/integrations/src/services/disbursement-schedule/e-cert-processing-steps/restriction-bypasses-resolution-step.ts:42]: Oh yes, I get it now. It is application based, so all good. Somehow I lost the way considered it to be student based.

**Comment by dheepak-aot** [comment]
Great work and clean implementation. Just wanted to check on adding an E2E scenario for accepting assessment with bypass. But it is optional as the path is validated.

File: sources/packages/backend/apps/queue-consumers/src/processors/schedulers/esdc-integration/ecert-integration/_tests_/ecert-full-time-process-integration.scheduler.e2e-spec.ts (hunk 10 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/apps/queue-consumers/src/processors/schedulers/esdc-integration/ecert-integration/_tests_/ecert-full-time-process-integration.scheduler.e2e-spec.ts]: Fixed.

**Review by sh16011993 — APPROVED** [review]
Nice work 👍  @andrewsignori-aot

**Comment by andrewsignori-aot** [comment]
> Great work and clean implementation. Just wanted to check on adding an E2E scenario for accepting assessment with bypass. But it is optional as the path is validated.

Added an E2E to accept assessment.
_Should allow NOA approval when the current assessment is blocked by a 'Stop part time disbursement' restriction but the application has a restriction bypass._

**Review by dheepak-aot — APPROVED** [review]
Thanks for making the changes. Looks good 👍
